### ISSUE TITLE
🐛 Bug Report: Proxy authorization header can't be defined as secret

### ISSUE THREAD

**Issue by dweber019** [backstage/backstage#17855]
### 📜 Description

I started to test https://github.com/backstage/backstage/tree/master/plugins/devtools. This plugin will expose a UI & endpoint which exposes the whole configuration.
There I saw that some sensible data is not defined as secrets. Therefor I opened some PR's to handle this

- https://github.com/backstage/backstage/pull/17852
- https://github.com/backstage/backstage/pull/17850
- https://github.com/RoadieHQ/roadie-backstage-plugins/pull/990

There could be more, these are only the ones I use and had an issue with.

The last issue was that I use a configuration like https://github.com/backstage/backstage/blob/master/app-config.yaml#L59-L62 with the proxy-backend:

To fix this I changed https://github.com/backstage/backstage/blob/master/plugins/proxy-backend/config.d.ts#L34 to something like
```
headers?: {
  /**
     * @visibility secret
     */
  Authorization?: string;
  [key: string]: string;
};

OR

headers?: {
  /**
     * @visibility secret
     */
  Authorization?: string;
};
```

### 👍 Expected behavior

The header value is treated as secret and will be replaced by https://github.com/backstage/backstage/blob/master/plugins/devtools-backend/src/api/DevToolsBackendApi.ts#L175

### 👎 Actual Behavior with Screenshots

The header is not treated as secret.

### 👟 Reproduction steps

1. Change the config.d.ts for the proxy backend as shown in the description
2. Go to the devtools page an check the output for the proxies

### 📃 Provide the context for the Bug.

I found that the reason for this behaviour are the slashes in the proxy configuration.
Doesn't work
```
'/jenkins/api': <========
    target: http://localhost:8080
    headers:
      Authorization: ${JENKINS_BASIC_AUTH_HEADER}
```
Works
```
'jenkins-api': <========
    target: http://localhost:8080
    headers:
      Authorization: ${JENKINS_BASIC_AUTH_HEADER}
```

As there are a lot of plugins using the authorization header in the proxy backend, this should probably be fixed. Or at least if not fixed a disclaimer should be added to the devtools plugin about this.
What do you think?

FYI @awanlin 

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, but I'm happy to collaborate on a PR with someone else

**Comment by awanlin** [comment]
Thanks for this logging this, this type of issue was always something I worried about. 

My thinking is that both approaches should be taken - setting the `@visibility secret` and updating the README with a disclaimer/note. The reality is that people will have custom things that we can't control so we should be clear to them about this.

I'll work on the README update, just need to think on how best to word it.

**Comment by afscrome** [comment]
Some other plugins pass API keys through other headers, so it's not just the Authorization header that can be affected.  E.g. [octopus](https://github.com/backstage/backstage/tree/master/plugins/octopus-deploy) uses `X-Octopus-ApiKey`

It's also worth noting that leaking the authorization header doesn't really reduce the security of the proxy (at least in the examples above).  When you set up credentials for the proxy, you are by definition creating a privilege escalation vulnerability. Any request made to the proxy path you configure will silently have the configured credentials injected.  Even when the credentials remain private, an attacker doesn't need to know the credentials as the proxy will helpfully add them of it's own accord.

See #17595 and the Proxy section of the [threat model](https://backstage.io/docs/overview/threat-model/#proxy) for more details.

**Comment by Shocktrooper** [comment]
Came here to say I ran across this in DevTools as well.  I wonder if there is a way to either override the schema or don't interpolate certain values in the DevTools UI.  In case it wasn't noted the database connection password is also in plaintext in the UI.

**Comment by dweber019** [comment]
@Shocktrooper You can overwrite the schema. For this you have to create a plugin with your own schema definition. But this doesn't work with the proxy backend as descriebed above.

@afscrome your right. We mitigated this by using read only, for any proxy based implementation and only allow GET. I think the bigger issue is that you don't have control what can be used with the proxy backend. Would be nice to have the permission framework implemented in the broxy backend so integrator could have control over the proxy on request level. I raised an issue once about this.

**Comment by Shocktrooper** [comment]
@dweber019 I found the schema config for the database connection and it is marked as secret but DevTools still shows it in plaintext which is interesting.  Do we know what potentially is happening.

- https://github.com/backstage/backstage/blob/85b3e4fcc17ce7950fcdf9ef8321b48982f11b00/packages/backend-common/config.d.ts#L71-L79

**Comment by awanlin** [comment]
Hi @Shocktrooper, I'm confused by this as it's properly hidden for us:

![image](https://github.com/backstage/backstage/assets/67169551/4e332848-4826-49d1-9826-40156343703e)

We are on `1.14.0` but we've been using this plugin for a long time now, since last September at least and I don't ever remember this database connection string being visible. There for sure were cases - like around the `auth` section that got fixed before this got merged. I totally missed the Proxy as we actually don't use it.

I'll try my best to fix things as I hear about them and I will be adding details about this in the README but I really the community to help me as it's hard to know all the possible ways this is being used.

Many thanks to @dweber019 for helping out, I really appreciate it! 🚀

**Comment by awanlin** [comment]
Ah, I think I know why it's not working as expected for you @Shocktrooper, are you using an object? I think that case is not working. Just loaded up my local and I don't use a connection string there and use the object and I can see the password, it's not getting masked.

This is what I mean by object:

```yaml
connection:
      host: 127.0.0.1
      user: fake_user
      password: fake_password
      port: 5432
```

**Comment by awanlin** [comment]
My apologies, we are using the Proxy, we added it back in December of 2022 for the Jira plugin. Totally forgot about that :(

Anyway, I've been working on a potential, fix/backup feature for this. What I was thinking is that you could provider an array of Property names that you want masked and when the config is being built up we would just mask any that matched. There could also be a default set. I'm just struggling with getting the property name as it's not included in the place I'm doing the current masking: 

https://github.com/backstage/backstage/blob/7b5425938284bb0f0ba1af57f09ccfdab81f0c03/plugins/devtools-backend/src/api/DevToolsBackendApi.ts#L189-L191

**Comment by Shocktrooper** [comment]
> Ah, I think I know why it's not working as expected for you @Shocktrooper, are you using an object? I think that case is not working. Just loaded up my local and I don't use a connection string there and use the object and I can see the password, it's not getting masked.
> 
> This is what I mean by object:
> 
> ```yaml
> connection:
>       host: 127.0.0.1
>       user: fake_user
>       password: fake_password
>       port: 5432
> ```

You are correct.  The documentation for setting up an external database uses the object notation instead of the connection string.  Potentially it should be added as a disclaimer to those docs.

Reference
- https://backstage.io/docs/tutorials/switching-sqlite-postgres/#add-postgresql-configuration

**Comment by Shocktrooper** [comment]
@Rugvip Are we good to bump this outside of the next backstage release? E.G. `yarn backstage-cli versions:bump --pattern '@backstage/config-loader'`

**Comment by awanlin** [comment]
@Shocktrooper I logged an issue for the database part #17955 and a potential PR fix #17956

**Comment by awanlin** [comment]
And PR #17957 adds a note about this to the DevTools README
